tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -9937,7 +9937,8 @@ FlattenedTypeIterator::FlattenedTypeIterator(SourceLocation loc, QualType type,
   m_source(source), m_draining(false), m_springLoaded(false), m_incompleteCount(0), m_typeDepth(0), m_loc(loc)
 {
   if (pushTrackerForType(type, nullptr)) {
-    considerLeaf();
+    while (!m_typeTrackers.empty() && !considerLeaf())
```

Comment:
The code here was assuming that one can recursively step into a struct to find its first (numerical) data member, but this didn't cover cases where the first member was an empty struct. In that case, we may need to step out and in the next field again.